nslsii/temperature_controllers/lakeshore.py

```diff
@@ -0,0 +1,126 @@
+        The keyword arguments passed to ``ophyd.Device``.
+    '''
+
+    def _set_fields(fields, device, prefix, **kwargs):
+        '''A function that allows for the components to be dynamically set.
+
+        Parameters
+        ----------
+        fields: list
+            A list of field identifiers to include, for ``_Temperature``
+            components an example is ['A','B','C','D'] and for ``_Control``
+            components an example is [1,2,3,4].
+        device: object
+            An ``ophyd.Device`` or ``ophyd.Signal`` object that shold be used
+            to create the ``ophyd.Component``s specifed in ``fields``.
+        prefix: str
+            The string to prefix to the field identifier in ``fields`` to
+            create the ``ophyd.Component`` ``suffix``.
+        **kwargs: dict
+            The kwargs that are to be passed into the call to ``device``.
+        '''
+        out_dict = OrderedDict()
+        for field in fields:
+            suffix = f'{prefix}{field}'
+            out_dict[f'{field}'] = (device, suffix, kwargs)
+        return out_dict
+
+    class _Temperature(Device):
+        '''A Device that is used to create the readback 'channel' components.
+
+        Parameters
+        ----------
+        *args: list
+            The arguments to pass down to ``ophyd.Device.__init__``
+        **kwargs: dict
+            The kwargs to be passed down to the ``ophyd.Device.__init__``
+        '''
+
+        T = Component(EpicsSignalRO, '}T-I')
+        T_celsius = Component(EpicsSignalRO, '}T:C-I')
+        V = Component(EpicsSignalRO, '}Val:Sens-I')
+        status = Component(EpicsSignalRO, '}T-Sts')
+        name = Component(EpicsSignal, '}T:Name-RB', write_pv='}T:Name-SP')
+        alarm_dict = {'high': (EpicsSignalRO, '}Alrm:High-Sts'),
+                      'low': (EpicsSignalRO, '}Alrm:Low-Sts')}
+        alarm = DynamicDeviceComponent(alarm_dict)
+        T_limit = Component(EpicsSignal, '}T:Lim-RB', write_pv='}T:Lim-SP')
+
+    class _Control(PVPositioner):
+        '''A sub-class used to define each of the temperature control channels.
+
+        Parameters
+        ----------
+        *args: list
+            The args to pass down to ``ophyd.PVPositioner.__init__``
+        **kwargs: dict
+            The kwargs to be passed down to the ``ophyd.PVPositioner.__init__``
+        '''
+        # PVPositioner required attributes
+        setpoint = Component(EpicsSignal, '}T-SP')
+        readback = Component(EpicsSignalRO, '}T-RB')
+        done = Component(EpicsSignalRO, '}Sts:Ramp-Sts')
+        # top level attributes
+        heater_range = Component(EpicsSignal, '}Val:Range-Sel')
+        heater_status = Component(EpicsSignalRO, '}Err:Htr-Sts')
+        mode = Component(EpicsSignal, '}Mode-Sel')
+        enable = Component(EpicsSignal, '}Enbl-Sel')
+        target_channel = Component(EpicsSignal, '}Out-Sel')
+        # ramp attributes
+        ramp_dict = {'enabled': (EpicsSignal, '}Enbl:Ramp-Sel'),
```

Comment:
This is going to be available on the final class, probbaly better to define this in-line with the DDC call it is used in.

Reply:
good call, done